### ISSUE TITLE
stop gradients for weights in tf.losses

### ISSUE THREAD

**Issue by ngc92** [tensorflow/tensorflow#15046]
In the case that the weights given to tf.losses.* depend in some way on the model parameters, 
the derivative of that loss also calculated with respect to the weights. 
(Stupid) minimal example:
```python
import tensorflow as tf
x = tf.constant(0)
w = tf.get_variable(name="W", shape=(), initializer=tf.zeros_initializer())
L = tf.losses.mean_squared_error(x, x, weights=w)
tf.train.AdamOptimizer().compute_gradients(L)
```
results in 
```
[(<tf.Tensor 'gradients/mean_squared_error/Mul_grad/tuple/control_dependency_1:0' shape=() dtype=float32>,
  <tf.Variable 'W:0' shape=() dtype=float32_ref>)]
```

I would expect the weights to be considered constant for the calculation of a loss. In case you agree with me, I can make a PR that adds `stop_gradient` around the weights parameter.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: N/A
- **TensorFlow installed from (source or binary)**: N/A
- **TensorFlow version (use command below)**: N/A
- **Python version**:  N/A
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: N/A

**Comment by facaiy** [comment]
Perhaps you'd better to use tf.constant for weight… instead of a variable.

**Comment by ngc92** [comment]
In my case the weights are not constant, but calculated in the model. The above example is only supposed to be a minimal example to demonstrate the problem.

**Comment by facaiy** [comment]
How about using `weights = weights * tf.ones((1))`?

**Comment by ngc92** [comment]
that wouldn't change anything. `weights + 0` is still differentiable w.r.t. `weights`. Also, just to clarify, my question is not how to stop `tf.losses.*` to derive at the weights, that would be a simple `tf.stop_gradient`, but whether it is conceptually correct that I would need to do so.

Having a weighted loss `L(x, w)` with values x ` and weights `w` both depending on a parameter vector `P`, which of the following would be considered correct
```
dL/dP = dL/dx * dx/dP
```
or
```
dL/dP = dL/dx * dx/dP + dL/dw * dw/dP
```

I would argue that the first one corresponds to the interpretation of  `w` as weights.

**Comment by facaiy** [comment]
Yup, the role of weight seems subtle when involved with variable. As said by you, we can pretend a variable as constant by tf.stop_gradients, so I'd prefer to the second one, which is more general.

**Comment by drpngx** [comment]
Right, I also disagree that we should stop the gradients. I understand where you're coming from, but I think this would be confusing non-transparent behavior for people. Also, once you stop the gradient within the loss it's not possible to go back to previous behavior.

**Comment by ngc92** [comment]
Then it would make sense to include this in the documentation, making this behaviour deliberate instead of a design accident.

**Comment by drpngx** [comment]
Sounds good. Feel free to send a PR
